docs/architecture/adr-015-namespace-id-size.md

```diff
@@ -0,0 +1,110 @@
+16 bytes (128 bits) | 0                | ~1.4432e-15        | ~1.4693e-9            | ~0.00147
+20 bytes (160 bits) | 0                | 0                  | 0                     | ~3.4205e-13
+32 bytes (256 bits) | 0                | 0                  | 0                     | 0
+
+> As a rule of thumb, a hash function with range of size N can hash on the order of sqrt(N) values before running into collisions.
+
+In other words
+
+Namespace ID size   | hash funciton range | can hash this many items before running into collision
+--------------------|---------------------|----------------------------------------------------------
+8 bytes (64 bits)   | 2^64                | 2^32 = ~4 billion items
+16 bytes (128 bits) | 2^128               | 2^64 = ~1 quintillion items
+20 bytes (160 bits) | 2^160               | 2^80 = ~1 septillion items
+32 bytes (256 bits) | 2^256               | 2^128 = ~3.4 quintillion items
+
+## Detailed Design
+
+1. What changes need to be made to celestia-app in order to support namespaces of a different length (e.g. 16 bytes)?
+    1. [done] Stop using the namespace ID defined by NMT
+    1. Increase `appconsts.NamespaceSize` to 16
+1. What changes need to be made to NMT in order to support namespaces of a different length (e.g. 16 bytes)?
+
+## Discussion Notes
+
+- Do we care about collisions created by an adversary?
+  - If so, an adversary can look at previously used namespace IDs and perform a woods attack on an existing namespace ID so increasing the namespace ID size doesn't resolve this threat.
+  - We should be careful in our documentation to not encourage users to assume that a randomly generated namespace ID is completely unique because:
+    - They could have generated a namespace ID without sufficient entropy
+    - An adversary can front-run a user's transaction and preemptively post to that namespace ID
+- For the non-adversarial use-case, we want to avoid users having to check if a random namespace has already been used.
+- Use case for larger namespace ID size: rollups may have multiple namespaces (e.g. Twitter) where a roll-up may give each user a namespace within a namespace range.
+- Is it possible to make the namespace ID a parameter, so that the namespace ID is a parameter to proof verification for roll-ups?
+  - Assumes that a block height may have a different namespace ID
+- There are talks in the Ethereum community about a potential address range increase.
+- 20 bytes gives us Ethereum address compatability so Ethereum addresses could be mapped to a Celestia namespace ID.
+- Currently namespace ID size = 8 bytes. Each intermediate node in the NMT is 8 (namespace ID) + 8 (namespace ID) + 32 (SHA256)= 48 bytes
+- If namespace ID size = 16 bytes. Each intermediate nodes in the NMT become: 16 (namespace ID) + 16 (namespace ID) + 32 (SHA256) = 64 bytes
+- Other option: increase size to 32 bytes with an optimization that reserves the first N bytes. The first N bytes wouldn't be sent over the wire.
```

Comment:
[Question]
I missed this point, what do you mean by:
 
> wouldn't be sent over the wire

They will be deterministic in some sense?

Reply:
The optimization would reserve the first N bytes to be unused initially. For example: if the total namespace ID size was 32 bytes, reserve the first 16 bytes as unusable initially. The first 16 bytes would be set to `0`. The last 16 bytes would contain a user's desired namespace.

When the share is sent across the network, a client could receive the first byte (the namespace version) and read that its of namespace version `0` so it can interpret the next 16 bytes as `0` and then read the next `16` bytes as the real namespace. In this optimization, the first 16 bytes of `0` don't need to be sent over the network. Does that make sense?

Reply:
Defintly, thanks :+1: :+1:

Reply:
The 0s also wouldn't be stored, right?

Reply:
If they aren't stored, then I don't understand where the namespace ID size actually represents 32 bytes. With this optimization, I expect the leading 16 bytes of `0`s to be stored in the share. 

Can you clarify where you don't think they would be stored @liamsi ?

Reply:
I don't think they need to be stored on disk, for example.

The only place where the 0s would be recognized is when computing the hash of intermediate nodes.

That being said, for this to solve the backwards-compatibility issue we're talking about, rollup software and NMT implementations would have to not assume that all intermediate nodes will have 0s at the start, in case we do ever increase the namespace size as use those preceding 0s...

Reply:
For example, assuming hashes are big endian, then when serializing the hashes over the wire or on disk, preceding 0s can be omitted, just like how Ethereum addresses that start up with 0s are more gas efficient

Reply:
> The only place where the 0s would be recognized is when computing the hash of intermediate nodes.

I think the 0s (or at least the share version) would also need to be recognized when creating a share commitment so that these two shares don't end up resulting in the same share commitment:

```go
shareA := []byte {
	0x0 // share version
	0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, // 16 bytes of unrecognized namespace
	0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, // 16 bytes of user specified namespace
	// remainder of share ommited for brevity
}
shareB := []byte {
	0x1 // share version
	0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, // 16 bytes of user specified namespace
	0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, // another 16 bytes of user specified namespace
	// remainder of share ommited for brevity
}
```

if that's the case, we're special casing the share version `0` to be interpreted as 16 bytes of `0`s in most instances where the namespace ID is used. In order to evaluate this optimization, I think we need to consider the code complexity and investigate how much bandwidth this decreases (presumably 16 bytes every time a namespace ID is sent over the network) but idk how much that is relative to total bandwidth requirements.

Reply:
Oh I see, yeah you can also make the 16 0s enforced by the share version, and if we increase the namespace size, introduce a new share version (which is opt-in by rollups thus it doesn't break backwards compatibility)

That being said, if a Merkle proof in an old share version contains a sibling for a new share version that has a longer namespace, that could break backwards compatibility, so you may need a bunch of padding shares ... (or I think the non-interactive defaults already solve this issue ?)

Reply:
> That being said, if a Merkle proof in an old share version contains a sibling for a new share version that has a longer namespace, that could break backwards compatibility

Agreed this does break backwards compatability. I don't expect non-interactive defaults to solve the issue. An NMT proof doesn't appear to contain the raw share of a sibling based on [these docs](https://github.com/celestiaorg/nmt/blob/1bc0bb0099e01b30e37ddb56642734ae875917cd/nmt.go#L150-L155) but `proof.nodes` contains namespaced hashes so if shareA and shareB are in the same data square and correspondingly same NMT, then inner nodes of the NMT for `namespaceVersion=0`  would be of size: 2 * 16 + 32 = 64 and inner nodes of the NMT for `namespaceVersion=1` would be of size: 2 * 32 + 32 = 96.

Reply:
In order for NMT proof verification to work, the original NMT proof verification logic would need to interpret `namespaceVersion=0` with 16 leading bytes of `0`s and to not perform such interpretation for `namespaceVersion=1` and above. It seems contrived to implement NMT proof verification in this way and it may in fact break if we made a different change when introducing `namespaceVersion=1`.

Preserving backwards compatibility of proofs for an NMT constructed with different size namespaces seems brittle in this scenario. Thoughts? @musalbas

Reply:
Yeah, you wouldn't have backwards compatibility if you have different namespace sizes for different versions anyways, because at least the root node might have a max or min node of a different size than older clients expect.

I think the only backwards compatible way is if all clients don't make any assumption about how many leading 0s there are, but have a consensus rule that requires leading 0s, like the Ethereum address optimization I mentioned here

>For example, assuming hashes are big endian, then when serializing the hashes over the wire or on disk, preceding 0s can be omitted, just like how Ethereum addresses that start with 0s are more gas efficient